servers/rendering_server.h

```diff
@@ -107,6 +107,7 @@ class RenderingServer : public Object {
 	virtual RID texture_proxy_create(RID p_base) = 0;
 
 	virtual void texture_2d_update(RID p_texture, const Ref<Image> &p_image, int p_layer = 0) = 0;
+	virtual void texture_2d_update_partial(RID p_texture, const Ref<Image> &p_sub_image, int p_dst_x, int p_dst_y, int p_mipmap = 0, int p_layer = 0) = 0;
```

Comment:
Performance: for cases where only a part of the source image are uploaded, it would be preferable if the method took a sub-region of the source image, rather than forcing to first instantiate and copy it from the source image just to be passed to this function.

Reply:
> it would be preferable if the method took a sub-region of the source image, rather than forcing to first instantiate and copy it from the source image just to be passed to this function.

I don't think there might be a big performance gap, and even if you do that, you're just calling Image.get_region() inside the method.

Reply:
But why is it necessary to even do that? It seems like a free thing to do in the first place 🤷

Reply:
Not sure, but I think the current approach is more in line with function names.
Moreover, Image.get_data() must be called internally anyway. This method produces a complete copy operation, and calling this function is expensive if the subimage is not obtained first.

Reply:
I would tend to agree with Zylann here, I would add arguments for source coordinates. Godot 3.x had the same and the currently unimplemented stub also has signature like that:

https://github.com/godotengine/godot/blob/578ca94ef6351cb6c31b1addf8b3c36577376bd4/drivers/gles3/storage/texture_storage.cpp#L1276-L1278



Implementing can be slightly more difficult as you also have to calculate stride, but basically it should be worth it in performance. My ultimate version (no need to implement) would probably take a PackedByteArray instead of Image, similar to high-performance mesh update methods like `mesh_surface_update_vertex_region`.

Also note that `Image.get_data()` returns a Vector, which is a copy-on-write container. It doesn't copy the contents until they are modified, which in this case is not needed.

Also, thanks for doing this! This will be a very useful method to many.

Reply:
> I would tend to agree with Zylann here, I would add arguments for source coordinates. Godot 3.x had the same and the currently unimplemented stub also has signature like that:
> 
> https://github.com/godotengine/godot/blob/578ca94ef6351cb6c31b1addf8b3c36577376bd4/drivers/gles3/storage/texture_storage.cpp#L1276-L1278
> 
> Implementing that can be slightly more difficult as you also have to calculate image stride and offsets, but it should be worth it in performance. My ultimate version (no need to implement) would probably take a PackedByteArray instead of Image, similar to high-performance mesh update methods like `mesh_surface_update_vertex_region`.
> 
> Also note that `Image.get_data()` returns a Vector, which is a copy-on-write container. It doesn't copy the contents until they are modified, which in this case is not needed.
> 
> Also, thanks for doing this! This will be a very useful method to many.

That makes sense, I've updated the api.

Reply:
I think the region should still be kept as small as possible. For example,  rgb8 is rendered as rgba8, a convert() will be generated, which is a very expensive operation.